site/pages/create-disruption.page.tsx

```diff
@@ -1,28 +1,144 @@
 import Select from "../components/form/Select";
+import Table from "../components/form/Table";
 import TextInput from "../components/form/TextInput";
 import TimeSelector from "../components/form/TimeSelector";
 import { BaseLayout } from "../components/layout/Layout";
 import { DISRUPTION_REASONS, COOKIES_DISRUPTION_INFO, COOKIES_DISRUPTION_ERRORS } from "../constants/index";
-import { PageState } from "../interfaces";
-import { createDisruptionSchema, Disruption } from "../schemas/create-disruption.schema";
-import { getStateUpdater } from "../utils/formUtils";
+import { ErrorInfo, PageState } from "../interfaces";
+import {
+    createDisruptionSchema,
+    validitySchemaRefined,
+    Validity,
+    validitySchema,
+} from "../schemas/create-disruption.schema";
+import { flattenZodErrors } from "../utils";
 
 const title = "Create Disruptions";
 const description = "Create Disruptions page for the Create Transport Disruptions Service";
 
-export interface DisruptionPageInputs extends Partial<Disruption> {}
+export interface DisruptionPageInputs extends Partial<z.infer<typeof createDisruptionSchema>> {}
 
 const CreateDisruption = (initialState: PageState<Partial<DisruptionPageInputs>>): ReactElement => {
     const [pageState, setDisruptionPageState] = useState<PageState<Partial<DisruptionPageInputs>>>(initialState);
+    const [validityButtonClicked, setValidityButtonClicked] = useState(false);
+    const initialValidity: Validity = {
+        id: "1",
+        disruptionStartDate: "",
+        disruptionEndDate: "",
+        disruptionStartTime: "",
+        disruptionEndTime: "",
+        disruptionNoEndDateTime: "",
+    };
+    const [validity, setValidity] = useState<Validity>(initialValidity);
+
+    const addValidity = (e: SyntheticEvent) => {
+        e.preventDefault();
+
+        const filteredValidity =
+            validity.disruptionNoEndDateTime === "true"
+                ? {
+                      ...validity,
+                      disruptionEndDate: "",
+                      disruptionEndTime: "",
+                  }
+                : validity;
+
+        const parsed = validitySchemaRefined.safeParse(filteredValidity);
+
+        if (parsed.success === false) {
+            setValidityButtonClicked(true);
+            setDisruptionPageState({
+                ...pageState,
+                errors: [
+                    ...pageState.errors.filter((err) => !Object.keys(validitySchema.shape).includes(err.id)),
+                    ...flattenZodErrors(parsed.error),
+                ],
+            });
+        } else {
+            setValidityButtonClicked(false);
+            setDisruptionPageState({
+                inputs: {
+                    ...pageState.inputs,
+                    validity: pageState.inputs.validity
+                        ? [
+                              ...pageState.inputs.validity,
+                              { ...filteredValidity, id: (pageState.inputs.validity.length + 1).toString() },
```

Comment:
This way of generating the id is leading to a bug since if you add 2 periods then remove the first one and add another one the 2 remaining periods will both have an id of "2", we either need to increment based on the id of the previous period in the array or we might be able to get rid of id and just use the index of the item in the array

Reply:
Just tested removing the id and it seems to work if we replace these functions with the following and then remove id from the zod schema:

```
    const addValidity = (e: SyntheticEvent) => {
        e.preventDefault();

        const filteredValidity =
            validity.disruptionNoEndDateTime === "true"
                ? {
                      ...validity,
                      disruptionEndDate: "",
                      disruptionEndTime: "",
                  }
                : validity;

        const parsed = validitySchemaRefined.safeParse(filteredValidity);

        if (parsed.success === false) {
            setDisruptionPageState({
                ...pageState,
                errors: [
                    ...pageState.errors.filter((err) => !Object.keys(validitySchema.shape).includes(err.id)),
                    ...flattenZodErrors(parsed.error),
                ],
            });
        } else {
            setDisruptionPageState({
                inputs: {
                    ...pageState.inputs,
                    validity: pageState.inputs.validity
                        ? [...pageState.inputs.validity, filteredValidity]
                        : [filteredValidity],
                },
                errors: [...pageState.errors.filter((err) => !Object.keys(validitySchema.shape).includes(err.id))],
            });

            setValidity(initialValidity);
        }
    };

    const removeValidity = (e: SyntheticEvent, index: number) => {
        e.preventDefault();
        if (pageState.inputs.validity) {
            const validity = [...pageState.inputs.validity];
            validity.splice(index, 1);

            setDisruptionPageState({
                inputs: {
                    ...pageState.inputs,
                    validity,
                },
                errors: pageState.errors,
            });
        }
    };

    const getValidityRows = () => {
        if (pageState.inputs.validity) {
            return pageState.inputs.validity.map((validity, i) => ({
                header: `Validity period ${i + 1}`,
                cells: [
                    validity.disruptionEndDate && validity.disruptionEndTime && !validity.disruptionNoEndDateTime
                        ? `${validity.disruptionStartDate} ${validity.disruptionStartTime} - ${validity.disruptionEndDate} ${validity.disruptionEndTime}`
                        : `${validity.disruptionStartDate} ${validity.disruptionStartTime} - No end date/time`,
                    <button
                        id={`${i + 1}`}
                        key={`remove-validity-period-${i + 1}`}
                        className="govuk-link"
                        onClick={(e) => removeValidity(e, i)}
                    >
                        Remove
                    </button>,
                ],
            }));
        }
        return [];
    };
```